### PR TITLE
Add cosine distance algorithm

### PR DESCRIPTION
**Pull request by richardnorth3**
This PR implements the cosine distance algorithm for both the DAAL and oneDAL frameworks, computing pairwise cosine distances between feature vectors. The DAAL implementation includes block-based, batch processing containers, and complete user documentation. The oneDAL implementation adds GPU offload with SYCL/DPC++ support through compute kernels for both CPU and GPU backends, comprehensive test cases for both CPU and GPU, bazel build integration, and updated documentation with mathematical specifications and API references.

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

### CODE REVIEW

File: cpp/daal/src/algorithms/cosdistance/cosdistance_batch_impl.i [hunk 1]
```diff
@@ -82,6 +82,26 @@ services::Status DistanceKernel<algorithmFPType, method, cpu>::compute(const siz
     }
 }
 
+template <typename algorithmFPType, Method method, CpuType cpu>
+services::Status DistanceKernel<algorithmFPType, method, cpu>::compute(const size_t na, const NumericTable * const a, const size_t nb,
```
Comment (ahuber21) [inline cpp/daal/src/algorithms/cosdistance/cosdistance_batch_impl.i]: What is the point of passing the tables as `const` and immediately `const_cast`ing it away? I think this should be avoided, even if it takes some effort. Removing `const` shouldn't be require for read-only access, no?

File: cpp/daal/src/algorithms/cosdistance/cosdistance_full_impl.i [hunk 2]
```diff
@@ -234,6 +234,89 @@ services::Status cosDistanceFull(const NumericTable * xTable, NumericTable * rTa
...
+
+            for (size_t i = 0; i < blockSize1; i++)
+            {
+                for (size_t j = 0; j < blockSize2; j++)
+                {
+                    algorithmFPType numerator = 0.0;
+                    algorithmFPType xNorm     = 0.0;
+                    algorithmFPType yNorm     = 0.0;
+
+                    for (size_t k = 0; k < p; k++)
```
Comment (ahuber21) [inline cpp/daal/src/algorithms/cosdistance/cosdistance_full_impl.i:293]: Did you check if the compiler is able to vectorize this loop?

File: cpp/daal/src/algorithms/cosdistance/cosdistance_full_impl.i [hunk 3]
```diff
@@ -234,6 +234,89 @@ services::Status cosDistanceFull(const NumericTable * xTable, NumericTable * rTa
...
+                    {
+                        numerator += x[i * p + k] * y[j * p + k];
+                        xNorm += x[i * p + k] * x[i * p + k];
+                        yNorm += y[j * p + k] * y[j * p + k];
+                    }
+
+                    algorithmFPType denominator = xNorm * yNorm;
+                    if (denominator > 0.0)
```
Comment (ahuber21) [inline cpp/daal/src/algorithms/cosdistance/cosdistance_full_impl.i:301]: It might be better to avoid the branch and use masking instead. @Vika-F recently mentioned such an example in a discussion.

File: cpp/daal/src/algorithms/cosdistance/cosdistance_full_impl.i [hunk 4]
```diff
@@ -234,6 +234,89 @@ services::Status cosDistanceFull(const NumericTable * xTable, NumericTable * rTa
     return safeStat.detach();
 }
 
+template <typename algorithmFPType, CpuType cpu>
+services::Status cosDistanceFull(const NumericTable * xTable, const NumericTable * yTable, NumericTable * rTable)
```
Comment (ahuber21) [inline cpp/daal/src/algorithms/cosdistance/cosdistance_full_impl.i:238]: General question about expected performance. We have quite some loop-nesting. What are expected values of `p`, `nBlocks1/2`, `blockSize1/2` that perform particularly good or bad?

I could use some more comments, maybe even an extended comment that sketches what's getting calculated and which element of the table are loaded when.

It looks like the inner-most loop is accessing memory sequentially, which would be good. But there's probably some connection between the dimension `p` and how efficient vectorization would be (the compiler can't know).
Maybe it's worth having a separate low-dim loop?

File: cpp/oneapi/dal/algo/cosine_distance.hpp [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include "oneapi/dal/algo/cosine_distance/compute.hpp"
```
Comment (ahuber21) [inline cpp/oneapi/dal/algo/cosine_distance.hpp]: Missing newline.

File: cpp/oneapi/dal/algo/cosine_distance/backend/cpu/compute_kernel.hpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+#ifdef ONEDAL_DATA_PARALLEL
+    void operator()(const dal::backend::context_cpu& ctx,
+                    const detail::descriptor_base<Task>& params,
+                    const table& x,
+                    const table& y,
+                    homogen_table& res) const;
+#endif
+};
+
+} // namespace oneapi::dal::cosine_distance::backend
```
Comment (ahuber21) [inline cpp/oneapi/dal/algo/cosine_distance/backend/cpu/compute_kernel.hpp]: Missing newline.

File: cpp/oneapi/dal/algo/cosine_distance/backend/cpu/compute_kernel_dense.cpp [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+                    homogen_table& res) const {
+        throw unimplemented(dal::detail::error_messages::method_not_implemented());
+    }
+#endif
+};
+
+template struct compute_kernel_cpu<float, method::dense, task::compute>;
+template struct compute_kernel_cpu<double, method::dense, task::compute>;
+
+} // namespace oneapi::dal::cosine_distance::backend
```
Comment (ahuber21) [inline cpp/oneapi/dal/algo/cosine_distance/backend/cpu/compute_kernel_dense.cpp]: \n

File: examples/oneapi/cpp/source/cosine_distance/cosine_distance_dense_batch.cpp [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    const auto x = dal::read<dal::table>(dal::csv::data_source{ data_file_name });
+    const auto y = dal::read<dal::table>(dal::csv::data_source{ data_file_name });
+    const auto distance_desc = dal::cosine_distance::descriptor{};
+
+    const auto result = dal::compute(distance_desc, x, y);
+
+    std::cout << "Values:\n" << result.get_values() << std::endl;
+
+    return 0;
+}
```
Comment (ahuber21) [inline examples/oneapi/cpp/source/cosine_distance/cosine_distance_dense_batch.cpp]: \n

(Did you use `clang-format`? It should have added those newlines. Maybe we need to check the config.)

File: examples/oneapi/dpc/source/cosine_distance/cosine_distance_dense_batch.cpp [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+int main(int argc, char const *argv[]) {
+    for (auto d : list_devices()) {
+        std::cout << "Running on " << d.get_platform().get_info<sycl::info::platform::name>()
+                  << ", " << d.get_info<sycl::info::device::name>() << "\n"
+                  << std::endl;
+        auto q = sycl::queue{ d };
+        run(q);
+    }
+    return 0;
+}
```
Comment (ahuber21) [inline examples/oneapi/dpc/source/cosine_distance/cosine_distance_dense_batch.cpp]: \n

File: cpp/daal/src/algorithms/cosdistance/cosdistance_batch_impl.i (hunk 1 above)
Comment (richardnorth3) [reply cpp/daal/src/algorithms/cosdistance/cosdistance_batch_impl.i]: This was included in the original implementation that I was following; I'm not sure myself why it was done in this manner, but I figure it may be a DAAL'ism that I haven't gotten a chance to understand yet. @Vika-F @icfaust, would you know the answer to this question?

Comment (Alexandr-Solovev) [reply cpp/daal/src/algorithms/cosdistance/cosdistance_batch_impl.i]: @ahuber21 Looks like a real bug which causes undefined behavior, but probably it makes sense to mark this as TODO and fix in next pr for all algos. @richardnorth3

File: examples/oneapi/cpp/source/cosine_distance/cosine_distance_dense_batch.cpp (hunk 8 above)
Comment (richardnorth3) [reply examples/oneapi/cpp/source/cosine_distance/cosine_distance_dense_batch.cpp]: `clang-format` would not add the last line for some reason. The config may have to be checked.

File: cpp/daal/src/algorithms/cosdistance/cosdistance_full_impl.i (hunk 4 above)
Comment (richardnorth3) [reply cpp/daal/src/algorithms/cosdistance/cosdistance_full_impl.i:238]: The goal of this PR is to add functionality. If the performance is poor, we can address the problem in a later PR, but it would be best to merge this PR as soon as possible.

File: cpp/daal/src/algorithms/cosdistance/cosdistance_full_impl.i (hunk 2 above)
Comment (Vika-F) [reply cpp/daal/src/algorithms/cosdistance/cosdistance_full_impl.i:293]: `#pragma omp simd reduction(+:numerator,xNorm,yNorm)` can help to be sure that the loop is vectorized.
But the syntax of the macros is going to be changed in ICX pragmas PR #3246

So, I'll update those during the merge conflicts resolution in #3246

File: cpp/daal/src/algorithms/cosdistance/cosdistance_batch_impl.i (hunk 1 above)
Comment (Vika-F) [reply cpp/daal/src/algorithms/cosdistance/cosdistance_batch_impl.i]: It looks like our technical debt.
As far as I remember, MKL wrappers like xgemm previously accepted only non-const data; so that's why those casts are here.
Currently they can be removed for xTable and yTable I think.

File: cpp/oneapi/dal/algo/cosine_distance/backend/gpu/compute_kernel_dense_dpc.cpp [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        ONEDAL_ASSERT(x.get_column_count() == y.get_column_count());
+
+        auto& queue = ctx.get_queue();
+        const auto x_nd = pr::table2ndarray<Float>(queue, x, sycl::usm::alloc::device);
+        const auto y_nd = pr::table2ndarray<Float>(queue, y, sycl::usm::alloc::device);
+
+        auto res_ptr = res.get_data<Float>();
+
+        // Temporary workaround until the table_builder approach is ready
+        auto res_nd = pr::ndarray<Float, 2>::wrap(const_cast<Float*>(res_ptr),
```
[the remark this answers is not in the document]
Comment (Vika-F) [reply cpp/oneapi/dal/algo/cosine_distance/backend/gpu/compute_kernel_dense_dpc.cpp:93]: Can this const_cast be avoided by replacing `auto` with `Float*`  and `get_data` with `get_mutable_data` in the previous line of code:

```
        Float* res_ptr = res.get_mutable_data<Float>();

        // Temporary workaround until the table_builder approach is ready
        auto res_nd = pr::ndarray<Float, 2>::wrap(res_ptr, { res.get_row_count(), res.get_column_count() });
```

[the remark this answers is not in the document]
Comment (richardnorth3) [reply cpp/oneapi/dal/algo/cosine_distance/backend/gpu/compute_kernel_dense_dpc.cpp:93]: As presently constructed, I do not believe this is possible. `res` is currently declared as a `homogen_table` and would have to use the array/ndarray class to leverage the `get_mutable_data` member function. Another alternative is whether we want to add this `get_mutable_data` functionality to the `homogen_table` class.

**Review by Vika-F — APPROVED** [review]
LGTM

File: cpp/daal/src/algorithms/cosdistance/cosdistance_batch_impl.i (hunk 1 above)
Comment (Vika-F) [reply cpp/daal/src/algorithms/cosdistance/cosdistance_batch_impl.i]: @richardnorth3
1) Can you please remove this compute() definition at all, and use the same approach as you used in correlation distance:
https://github.com/uxlfoundation/oneDAL/blob/main/cpp/daal/src/algorithms/cordistance/cordistance_batch_impl.i#L69
2) const_casting for xTable and yTable is not needed here (and in correlation distance) as cosDistanceFull() function accepts const pointers.

File: cpp/daal/src/algorithms/cosdistance/cosdistance_batch_impl.i [hunk 11]
```diff
@@ -63,7 +63,19 @@ services::Status DistanceKernel<algorithmFPType, method, cpu>::compute(const siz
...
     if (isFull<algorithmFPType, cpu>(rLayout))
     {
-        return cosDistanceFull<algorithmFPType, cpu>(xTable, rTable);
+        if (na == 1)
+        {
+            return cosDistanceFull<algorithmFPType, cpu>(xTable, rTable);
+        }
+        else if (na == 2)
+        {
+            NumericTable * yTable = const_cast<NumericTable *>(a[1]); /* y Input data */
```
Comment (Vika-F) [inline cpp/daal/src/algorithms/cosdistance/cosdistance_batch_impl.i:72]:
```suggestion
            const NumericTable * yTable = a[1]; /* y Input data */
```

Comment (Vika-F) [reply cpp/daal/src/algorithms/cosdistance/cosdistance_batch_impl.i:72]: Similar changes can be done in the lines 60, 61:
```
const NumericTable * xTable = a[0]; /* Input data */
NumericTable * rTable       = r[0]; /* Output data */
```

**Review by ahuber21 — APPROVED** [review]
Thanks for taking care of the issues as good as possible. @Vika-F let's keep an eye on this and aim for gradual improvement with future PRs.

**Comment by richardnorth3** [comment]
Thank you @Vika-F @ahuber21 for the help!
